internal/extension/reconciler.go

```diff
@@ -4,12 +4,94 @@ import (
 	"context"
 	"sync"
 	"sync/atomic"
+	"time"
 
 	"github.com/kuadrant/policy-machinery/controller"
 	"github.com/kuadrant/policy-machinery/machinery"
 )
 
-var DAG *atomic.Pointer[StateAwareDAG]
+// nilGuardedPointer is an atomic pointer that provides blocking behavior
+// until the pointer is set to a non-nil value.
+type nilGuardedPointer[T any] struct {
+	ptr  atomic.Pointer[T]
+	mu   sync.Mutex
+	cond *sync.Cond
+}
+
+// newNilGuardedPointer creates a new nilGuardedPointer.
+func newNilGuardedPointer[T any]() *nilGuardedPointer[T] {
+	cgp := &nilGuardedPointer[T]{}
+	cgp.cond = sync.NewCond(&cgp.mu)
+	return cgp
+}
+
+// set sets the pointer to x and signals any goroutines waiting for a non-nil value.
+func (c *nilGuardedPointer[T]) set(x T) {
+	c.ptr.Store(&x)
```

Comment:
`get()` returns `*T`. Should `set(x)` expect `*T` as well?

```suggestion
func (c *nilGuardedPointer[T]) set(x *T) {
	c.ptr.Store(x)
```

Reply:
Well, no, because I don't want it settable to `nil` again, the idea is _only_ support `nil` -> `T` transitions. `get` (and the timeout'ing version) have to return a pointer, as they both could still return `nil`, `getWait` can only ever return a `T`. Now it's probably worth noting this used to store (and guard) a `StateAwareDAG`, so this only "copies" two pointers around